lib/ansible/modules/crypto/openssh_keypair.py

```diff
@@ -0,0 +1,327 @@
+
+        if self.type == 'dsa':
+            self.size = 1024 if self.size is None else self.size
+            if self.size != 1024:
+                module.fail_json(msg='''DSA keys must be exactly 1024 bits as specified by FIPS 186-2.''')
+
+        if self.type == 'ecdsa':
+            self.size = 256 if self.size is None else self.size
+            if self.size not in (256, 384, 521):
+                module.fail_json(msg='''For ECDSA keys, size determines the key length by selecting from
+                                     one of three elliptic curve sizes: 256, 384 or 521 bits.
+                                     Attempting to use bit lengths other than these three values for
+                                     ECDSA keys will cause this module to fail.''')
+        if self.type == 'ed25519':
+            self.size = 256
+
+    def generate(self, module):
+        # generate a keypair
+        if not self.check(module, perms_required=False) or self.force:
+            args = [
+                'ssh-keygen',
+                '-q',
+                '-b', str(self.size),
+                '-t', self.type,
+                '-f', self.path,
+            ]
+
+            if self.passphrase:
+                args.extend(['-N', self.passphrase])
+            else:
+                args.extend(['-N', ''])
+
+            if self.comment:
+                args.extend(['-C', self.comment])
+            else:
+                args.extend(['-C', ""])
+
+            try:
+                subprocess.call(args)
+                self.changed = True
+            except subprocess.CalledProcessError as e:
+                self.remove()
+                module.fail_json(msg="%s" % to_native(e))
+
+        proc = Popen(['ssh-keygen', '-lf', self.path], stdout=PIPE)
+        self.fingerprint = proc.communicate()[0].split()
+
+        file_args = module.load_file_common_arguments(module.params)
+        if module.set_fs_attributes_if_different(file_args, False):
+            self.changed = True
+
+    def check(self, module, perms_required=True):
+
+        # check if the key is correct
+        def _check_state():
+            return os.path.exists(self.path)
+
+        if _check_state():
+            proc = Popen(['ssh-keygen', '-lf', self.path], stdout=PIPE)
+            fingerprint = proc.communicate()[0].split()
+            keysize = int(fingerprint[0])
+            keytype = fingerprint[-1][1:-1].lower()
+        else:
+            return False
+
+        def _check_perms(module):
+            file_args = module.load_file_common_arguments(module.params)
+            return not module.set_fs_attributes_if_different(file_args, False)
+
+        def _check_type():
+            return self.type == keytype.decode("utf-8")
+
+        def _check_passphrase():
+            self.passphrase = '' if self.passphrase is None else self.passphrase
+            proc = Popen(['ssh-keygen', '-q', '-y', '-P', self.passphrase, '-f', self.path], stdout=PIPE)
```

Comment:
Communicating the passphrase as an argument to `ssh-keygen` may make it show up in process lists. This is not an acceptable leak. I did not see an easy way to make it listen on `stdin` for the password, though...

```
>>> import subprocess
>>> subprocess.Popen(['/usr/bin/sleep', '100'])
```

![the arguments to sleep are visible in htop](https://user-images.githubusercontent.com/974662/46491261-80be8d80-c80a-11e8-9cf1-f64663e96fca.png)

Reply:
That's indeed quite a problem. I don't see how this can be solved in a good way (except to essentially rewrite ssh-keygen in Python).

Related, but unrelated: to execute commands from an Ansible module, there's `module.run_command()`. There's also `module.get_bin_path()` to get hold of the path of the binary. Search the modules for examples.

Reply:
`pexpect` could be used to provide the password to the subprocess on stdin.

Reply:
i solved it by using the SSH_ASKPASS envoirment variable. Providing it through stdin jus isn't reliable enough.